pkg/proxy/metrics/metrics.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	compbasemetrics "k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+// used in the middleware below to capture http status code.
+type statusCatcher struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func (s *statusCatcher) WriteHeader(statusCode int) {
+	s.statusCode = statusCode
+	s.ResponseWriter.WriteHeader(statusCode)
+}
+
+// WithRequestCounter counts the number of requests that have passed through
+// the handler it wraps.
+func WithRequestCounter(delegate http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		sc := &statusCatcher{
+			ResponseWriter: w,
+		}
+
+		delegate.ServeHTTP(sc, r)
+
+		code := strconv.Itoa(sc.statusCode)
+		verb := strings.ToUpper(r.Method)
+		requestCounter.WithContext(r.Context()).WithLabelValues(verb, code).Inc()
+	})
+}
+
+// WithLatencyTracking tracks the number of seconds it took the wrapped handler
+// to complete.
+func WithLatencyTracking(delegate http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		sc := &statusCatcher{
+			ResponseWriter: w,
+		}
+
+		start := time.Now()
+		delegate.ServeHTTP(sc, r)
+		elapsedSeconds := time.Since(start).Seconds()
```

Comment:
Depending on how these get layered, you're going to count the time taken to observe metrics at the lower level in this. Might want to think of a factoring that only ends up adding one handler into the chain to avoid that, and any duplicate processing we're going to do (e.g. extracting cluster name, shards, etc from the request)

Reply:
> Depending on how these get layered, you're going to count the time taken to observe metrics at the lower level in this

I think that's okay unless you feel strongly otherwise. "One wrapper per metric" seems to be a pattern [prometheus uses](https://github.com/prometheus/client_golang/blob/main/prometheus/promhttp/instrument_server.go).

> and any duplicate processing we're going to do (e.g. extracting cluster name, shards, etc from the request)

I plan to open a followup PR that will extract the shard info near the top of the chain and attach it to the context so metrics and the shard reverse proxies can use it.